read_edn.py

```diff
@@ -1,18 +1,214 @@
+    latest_founded_date = get_founded_date(latest_founded_asset)
+    print(
+        f"Earliest backtest date is {latest_founded_date} (when {latest_founded_asset} was founded)")
+
+
+def get_founded_date(ticker: str) -> datetime.date:
+    # TODO: look values up (I have this in the ETF DB)
+    return datetime.date(2017, 7, 1)
+
+
+def collect_allocateable_assets(node) -> set[str]:
+    s = set()
+    if node[':step'] == ":asset":
+        s.add(node[':ticker'])
+        return s  # this is a leaf node, can end here
+
+    for child in node[":children"] if ":children" in node else []:
+        s.update(collect_allocateable_assets(child))
+    return s
+
+
+def collect_if_referenced_assets(node) -> set[str]:
+    """
+    Collects tickers referenced by if-conditions
+    """
+    s = set()
+    if node[':step'] == ":asset":
+        # s.add(node[':ticker'])
+        return s  # this is a leaf node, can end here
+    elif node[':step'] == ":if-child" and not node[":is-else-condition?"]:
+        if type(node[':lhs-val']) == str:
+            s.add(node[':lhs-val'])
+        if not node.get(':rhs-fixed-value?', type(node[':rhs-val']) != str):
+            s.add(node[':rhs-val'])
+
+    for child in node[":children"] if ":children" in node else []:
+        s.update(collect_if_referenced_assets(child))
+    return s
+
+
+def collect_referenced_assets(node) -> set[str]:
+    s = set()
+    s.update(collect_if_referenced_assets(node))
+    s.update(collect_allocateable_assets(node))
+    return s
+
+
+def print_children(node, depth=0):
+    """
+    Recursively visits every child node (depth-first)
+    and pretty-prints it out.
+    """
+    # Every node has ":id", maybe ":children", and ":step"
+    node_type = node[":step"]
+    if node_type == ":root":
+        # ':name': str
+        # ':description': str
+        # ':rebalance': TODO
+        print("  " * depth, node[":name"])
+    elif node_type == ":wt-cash-equal":
+        # ':window-days': Optional[int] TODO: what does it mean?
+        print("  " * depth, "Weight equally:")
+
+    elif node_type == ":if":
+        # All :children are definitely if-child
+
+        print("  " * depth, "if")
+    elif node_type == ":if-child":
+        # ':is-else-condition?'
+
+        if node[":is-else-condition?"]:
+            print("  " * depth, "else")
+        else:
+            # ':rhs-fn'
+            # ':rhs-fixed-value?'
+            # ':rhs-window-days'
+            # ':rhs-val'
+
+            # ':lhs-fn'
+            # ':lhs-window-days'
+            # ':lhs-val'
+
+            # ':comparator'
+
+            # NOTE: parameter if :rhs-fixed-value
+            def pretty_fn(fn_string: str) -> str:
+                if fn_string == ":relative-strength-index":
+                    return "RSI"
+                return fn_string
+
+            def pretty_comparison(comparator_string: str) -> str:
+                if comparator_string == ":lte":
+                    return "≤"
```

Comment:
Can we return "<=" instead, since we can then this for python later.
Same for below...

Reply:
Sure, will be in next PR. https://github.com/jamesfulford/compose_symphony_parser/commit/5db487514ea2df4db1f6c7cc135bc0aa6f07f681